internal/contract/infrastructure_cluster.go

```diff
@@ -61,6 +61,79 @@ func (c *InfrastructureClusterControlPlaneEndpoint) Port() *Int64 {
+func (c *InfrastructureClusterContract) FailureReason() *String {
+	return &String{
+		path: []string{"status", "failureReason"},
+	}
+}
+
+// FailureMessage provides access to the status.failureMessage field in an InfrastructureCluster object. Note that this field is optional.
+func (c *InfrastructureClusterContract) FailureMessage() *String {
+	return &String{
+		path: []string{"status", "failureMessage"},
+	}
+}
+
+// Option 1 for FailureDomains:
+// Pros: Provides granularity for accessing map[string]FailureDomainSpec using intermediate structs, all `path` return
+// types are wrappers for primatives instead of custom structs.
+// Cons: Accessing the map[string]FailureDomainSpec requires knowing the keys ahead of time.
+
+// FailureDomains provides access to the status.failureDomains field in an InfrastructureCluster object. Note that this field is optional.
+func (c *InfrastructureClusterContract) FailureDomains() *InfrastructureClusterFailureDomains {
+	return &InfrastructureClusterFailureDomains{}
+}
+
+// InfrastructureClusterFailureDomains provides a helper struct for working with FailureDomains
+// in an InfrastructureCluster object.
+type InfrastructureClusterFailureDomains struct{}
+
+// FailureDomainSpec provides access to the status.failureDomains.<name> field in an InfrastructureCluster object. Note that this field is optional.
+func (c *InfrastructureClusterFailureDomains) FailureDomainSpec(name string) *InfrastructureClusterFailureDomainSpec {
+	return &InfrastructureClusterFailureDomainSpec{
+		Name: name,
+	}
+}
+
+// InfrastructureClusterFailureDomainSpec provides a helper struct for working with FailureDomainSpec
+// in an InfrastructureCluster object.
+type InfrastructureClusterFailureDomainSpec struct {
+	Name string
+}
+
+// ControlPlane provides access to the status.failureDomains.<name>.controlPlane field in an InfrastructureCluster object. Note that this field is optional.
+func (c *InfrastructureClusterFailureDomainSpec) ControlPlane() *Bool {
+	return &Bool{
+		path: []string{"status", "failureDomains", c.Name, "controlPlane"},
+	}
+}
+
+// Attributes provides access to the status.failureDomains.<name>.attributes field in an InfrastructureCluster object. Note that this field is optional.
+func (c *InfrastructureClusterFailureDomainSpec) Attributes() *StringMap {
+	return &StringMap{
+		path: []string{"status", "failureDomains", c.Name, "attributes"},
+	}
+}
+
+// Option 2 for FailureDomains:
+// Pros: Directly returns the map[string]FailureDomainSpec, allowing users to determine the keys at runtime.
+// Cons: Requires defining a new intermediate struct and doesn't document the subfields within a FailureDomainSpec.
+
+// FailureDomains provides access to the status.failureDomains field in an InfrastructureCluster object. Note that this field is optional.
+func (c *InfrastructureClusterContract) FailureDomainsAlt() *FailureDomains {
```

Comment:
Here's the second way we could implement it such that we just return the `map[string]FailureDomainSpec` directly. While this method doesn't require knowing the keys ahead of time, it does require defining a new `FailureDomains` type with the getter and setter.

Reply:
I would prefer this option. I think both options require to define a type (above it was `InfrastructureClusterFailureDomains` here it is `FailureDomains`).

Feels more straightforward to me to just return the entire clusterv1.FailureDomains structure

Reply:
@fabriziopandini WDYT?